org.eclipse.ice.tasks/src/main/java/org/eclipse/ice/tasks/Task.java

```diff
@@ -0,0 +1,270 @@
+		smExecutingAction.setTaskAction(action.get());
+		// Execute the action. Note that there is a transition event tied to the state
+		// event for execution. The transition event changes the state to EXECUTING
+		// while the state event makes some more informed decisions.
+		stateMachine.sendEvent(TaskTransitionEvents.EXECUTION_TRIGGERED);
+		logger.info("Task execution initiated.");
+		return getState();
+	}
+
+	@Override
+	public TaskState cancel() {
+		// TODO Auto-generated method stub
+		return null;
+	}
+
+	@Override
+	public TaskState getState() {
+		return stateData.get().getTaskState();
+	}
+
+	@Override
+	public TaskStateData getTaskStateData() {
+		return (TaskStateData) stateData.get().clone();
+	}
+
+	/**
+	 * This operation builds the state machine used to manage states and transitions
+	 * for Tasks.
+	 * 
+	 * @throws TaskException
+	 * 
+	 * @throws Exception     thrown in the state machine can not be correctly
+	 *                       assembled.
+	 */
+	private void buildStateMachine() throws TaskException {
+
+		try {
+			// Setup the state machine builder
+			Builder<TaskState, TaskTransitionEvents> builder = new StateMachineBuilder.Builder<>();
+
+			// Go into the intial state and add the others states to the set
+			builder.configureStates().withStates()
+					.initial(TaskState.INITIALIZED, new InitializedStateAction<>(stateData.get()))
+					.states(EnumSet.allOf(TaskState.class));
+
+			// Configure the transitions for action data and actions.
+			builder.configureTransitions().withExternal().source(TaskState.INITIALIZED).target(TaskState.WAITING)
+					.event(TaskTransitionEvents.ACTION_DATA_SET).action(new WaitingEventAction<>(stateData.get()));
+			builder.configureTransitions().withExternal().source(TaskState.INITIALIZED).target(TaskState.WAITING)
+					.event(TaskTransitionEvents.ACTION_SET).action(new WaitingEventAction<>(stateData.get()));
+
+			// Once all the info is set, the state machine can transition to READY
+			builder.configureTransitions().withExternal().source(TaskState.WAITING).target(TaskState.READY)
+					.event(TaskTransitionEvents.ALL_INFO_SET).action(new ReadyEventAction<>(stateData.get()));
+
+			// A call to execute() triggers execution. This requires a transition action...
+			builder.configureTransitions().withExternal().source(TaskState.READY).target(TaskState.EXECUTING)
+					.event(TaskTransitionEvents.EXECUTION_TRIGGERED)
+					.action(new ExecutingEventAction<>(stateData.get()));
+			// ...and a state action. This action requires all of the data.
+			builder.configureStates().withStates().state(TaskState.EXECUTING, smExecutingAction);
+
+			// If the execution finished as expected, there is a transition event for that
+			// to finalize the state.
+			builder.configureTransitions().withExternal().source(TaskState.EXECUTING).target(TaskState.FINISHED)
+					.event(TaskTransitionEvents.EXECUTION_FINISHED).action(new FinishedEventAction<>(stateData.get()));
```

Comment:
In the event that the execution did not finish as expected, should it enter the `TaskState.FAILED` state? Does that transition need to be defined here also?

Reply:
With respect to differences...
1. See the code. This uses a new state machine library instead of the one I wrote from scratch. Also much cleaner.
2. See my thesis. The workflow model matches that with some corrections.

Reply:
With respect to entering TaskState.FAILED, yes it should. However, that is a state action and what you are looking at here is a transition action. So the movement to TaskState.Failed has to be triggered from within the Action itself and isn't listed here.